internal/clients/team4/client.go

```diff
@@ -272,7 +285,47 @@ func (c *client) StartOfTurn() {
 
 func (c *client) updateTrustFromSavedHistory() {
 	if c.savedHistory.updated {
+		newInfo := c.savedHistory.getNewInfo()
 
+		if len(newInfo) > 0 {
+			var truthfulnessSum float64
+
+			for _, history := range newInfo {
+				truthfulnessSum += history.TruthfulRatio
+			}
+			averageTruthfulness := truthfulnessSum / float64(len(newInfo))
+
+			for clientID, history := range newInfo {
+				truthfulness := history.TruthfulRatio
+
+				c.trustMatrix.ChangeClientTrust(clientID, c.internalParam.historyWeight*(truthfulness-averageTruthfulness)) //potentially add * historyWeight to scale the update
+
+				if floatEqual(truthfulness, 1) { //bonus for being fully truthful
+					c.trustMatrix.ChangeClientTrust(clientID, c.internalParam.historyFullTruthfulnessBonus)
+				}
+			}
+		}
+		c.savedHistory.updated = false
+	}
+}
+
+func (c *client) updateTrustMonitoring(data map[shared.CommunicationFieldName]shared.CommunicationContent) {
+	if roleMonitored, ok := data[shared.RoleMonitored]; ok {
+		if roleMonitored.T == shared.CommunicationIIGORole {
+			if monitoringResult, ok := data[shared.MonitoringResult]; ok {
+				if monitoringResult.T == shared.CommunicationBool {
+					roleID := c.getRole(roleMonitored.IIGORoleData)
+
+					if monitoringResult.BooleanData {
+						// Monitored role was truthful
+						c.trustMatrix.ChangeClientTrust(roleID, c.internalParam.monitoringWeight*c.internalParam.monitoringResultChange) // config?
+					} else {
+						// Monitored role cheated
+						c.trustMatrix.ChangeClientTrust(roleID, -c.internalParam.monitoringWeight*c.internalParam.monitoringResultChange)
+					}
```

Comment:
Slightly experimental, but could you have something like a bool to +/-, so true evaluate to 1.0 and false evaluate to -1.0?

Reply:
I think this is more readable